pvlib/irradiance.py

```diff
@@ -1198,6 +1198,44 @@ def perez(surface_tilt, surface_azimuth, dhi, dni, dni_extra,
         return sky_diffuse
 
 
+def clearsky_index(ghi, clearsky_ghi, max_clearsky_index=2.0):
+    """
+    Calculate the clearsky index.
+
+    The clearsky index is the ratio of global to clearsky global irradiance.
+    Negative and non-finite clearsky index values will be truncated to zero.
+
+    Parameters
+    ----------
+    ghi : numeric
+        Global horizontal irradiance in W/m^2.
+
+    clearsky_ghi : numeric
+        Modeled clearsky GHI
+
+    max_clearsky_index : numeric, default 2.0
+        Maximum value of the clearsky index. The default, 2.0, allows
+        for over-irradiance events typically seen in sub-hourly data.
+
+    Returns
+    -------
+    clearsky_index : numeric
+        Clearsky index
+    """
+    clearsky_index = ghi / clearsky_ghi
```

Comment:
Good catch thanks. I think in this function we can get by with testing if the input clearsky is small (< .001?).  I suggest we discuss the proposed division function elsewhere.

Reply:
I agree that the cutoff exactly at zero might not be sufficient in some cases. I've probably missed some discussions on this, but for example it is unclear to me how the perez model is supposed to behave as both `dni` and `dhi` irradiance get very close to zero: https://github.com/pvlib/pvlib-python/blob/master/pvlib/irradiance.py#L1126

Reply:
Reprise of #433 .  For now let's have the clearsky index function calculate the ratio of GHI to clearsky GHI, with polite treatment of Inf, 0 division and nan, and not assume how very low GHI values should be handled, especially output of clearsky models.

Reply:
I don't think we should assume how very low values of `ghi` should be treated, but I think we can and should make some assumptions about low values of `clearsky_ghi`. For nighttime values, we can expect ratios such as `1 / 0` and I hope we agree that this should return 0 instead of `max_clearsky_index`. Near sunrise/sunset, we could occasionally see ratios such as `1 / 0.001`. I think this should return 0, but I won't object if this returns `max_clearsky_index`.

Reply:
I agree with returning 0 for nighttime as defined by the clearsky GHI. Probably should infer nighttime based on solar angle rather than clearsky GHI but let's do that in a different PR.

I would prefer returning `max_clearsky_index` for large ratios to get this PR to closure. I'd prefer to return 0 for low values of clearsky GHI, but that is asking for another kwarg, e.g., `min_clearksy_ghi` for this function.

Reply:
Ok. Then this PR only needs to be modified to preserve `nan` inputs. Everything else is kicked down the road.

Reply:
I've seen elsewhere the idiom of multiplying a final result by eg `(ghi / ghi)` to preserve `nan`s in the input.  Is that approach acceptable here as well?  If so, I will add that in and update the tests after work this evening.

Reply:
@markcampanelli and @anomam IMHO [`divide_without_warning()`](#discussion_r) seems like a lot of overhead for something that it might be possible to solve using with [Python `warnings`](https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings) and maybe `np.where()`. I agree this should be considered in a separate PR that handles these types of limits, `NaN` returns, warnings, etc. in a simple, stable, and efficient way that we mostly agree on.

Reply:
> I've seen elsewhere the idiom of multiplying a final result by eg (ghi / ghi) to preserve nans in the input. Is that approach acceptable here as well? 

No objection if commented appropriately. It may or may not be more efficient that other approaches, but let's not worry too much about that unless someone complains.

Reply:
Multiplying by `(ghi/ghi)` actually introduced new nan values when ghi=0, so I opted to just fill `nan` back into the array where appropriate with the same type-preserving methods as before.